test/e2e/apps/daemon_set.go

```diff
@@ -746,3 +780,189 @@ func waitFailedDaemonPodDeleted(c clientset.Interface, pod *v1.Pod) func() (bool
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		return nil, fmt.Errorf("failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	return pods, nil
+}
+
+func checkDaemonSetPodsOrphaned(podClient corev1typed.PodInterface, dsName string) error {
+	label := map[string]string{daemonsetNameLabel: dsName}
+	if err := wait.PollImmediate(dsRetryPeriod, dsRetryTimeout, func() (bool, error) {
+		pods, err := getPods(podClient, label)
+		if err != nil {
+			return false, err
+		}
+
+		for _, pod := range pods.Items {
+			// This pod is orphaned only when its controllerRef is nil
+			if controllerRef := metav1.GetControllerOf(&pod); controllerRef != nil {
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("daemonset pods are not orphaned: %v", err)
+	}
+	return nil
+}
+
+func listDaemonSetHistories(controllerRevisionClient appstyped.ControllerRevisionInterface, labelMap map[string]string) (*apps.ControllerRevisionList, error) {
+	selector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: selector.String()}
+	historyList, err := controllerRevisionClient.List(options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list daemonset histories: %v", err)
+	}
+	return historyList, nil
+}
+
+func checkDaemonSetHistoriesOrphaned(controllerRevisionClient appstyped.ControllerRevisionInterface, dsName string) error {
+	label := map[string]string{daemonsetNameLabel: dsName}
+	if err := wait.PollImmediate(dsRetryPeriod, dsRetryTimeout, func() (bool, error) {
+		histories, err := listDaemonSetHistories(controllerRevisionClient, label)
+		if err != nil {
+			return false, err
+		}
+
+		for _, history := range histories.Items {
+			// This history is orphaned only when its controllerRef is nil
+			if controllerRef := metav1.GetControllerOf(&history); controllerRef != nil {
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("daemonset histories are not orphaned: %v ", err)
+	}
+	return nil
+}
+
+func waitDaemonSetAdoption(
+	podClient corev1typed.PodInterface,
+	controllerRevisionClient appstyped.ControllerRevisionInterface,
+	ds *apps.DaemonSet,
+	podNamePrefix string) error {
+	if err := checkDaemonSetPodsAdopted(podClient, ds.UID); err != nil {
+		return err
+	}
+	if err := checkDaemonSetHistoriesAdopted(controllerRevisionClient, ds); err != nil {
+		return err
+	}
+	// Ensure no pod is re-created by checking their names
+	if err := checkDaemonSetPodsName(podClient, ds.Name, podNamePrefix); err != nil {
+		return err
+	}
+	return nil
+}
+
+func checkDaemonSetPodsAdopted(podClient corev1typed.PodInterface, dsUID types.UID) error {
+	if err := wait.PollImmediate(dsRetryPeriod, dsRetryTimeout, func() (bool, error) {
+		pods, err := podClient.List(metav1.ListOptions{LabelSelector: labels.Everything().String()})
```

Comment:
Please use the function that lists the DaemonSet's pods

Reply:
I suggest retaining this line. Based on your other feedback, I changed `listPods` function arguments to `listPods(..., label map[string]string)`, in which `label` is `ds.Spec.Selector.MatchLabels`.

There is no easy way to convert `labels.Everything()` of type `labels.Selector` to `map[string]string` format. A fix is to change `listPods` function arguments to `listPods(..., labelSelector labels.Selector)`, which I don't prefer.